File: 2441. Largest Positive Integer That Exists With Its Negative.js
//2441. Largest Positive Integer That Exists With Its Negative
nums = [-1,2,-3,3,-2]
/**
 * @param {number[]} nums
 * @return {number}
 */
//SOL 1 | OG
var findMaxK = function(nums) {
  const a = nums.filter(num => num > 0)
  const b = a.filter(num => nums.includes(-num))
  console.log(b)
  return b.length === 0 ? -1 : Math.max(...b);
};
console.log(findMaxK(nums))

//SOL 2 | OPTIMIZED
var findMaxK = function(nums) {
  let setNums = new Set(nums); 
  let lp = -1; 
  for (let num of setNums) {
    if(setNums.has(-num) && num > 0 && num > lp)  lp = num;
  }
  return lp; 
};
console.log(findMaxK(nums))

//SOL 2.1 | COMMENTS
var findMaxK = function(nums) {
  let setNums = new Set(nums); // Create a set to store unique numbers from the input array.
  let lp = -1; // Initialize the largest positive number found so far to -1.

  // Iterate through each number in the set of unique numbers.
  for (let num of setNums) {
    // Check if the set contains the negative of the current number, and if the current number is positive and larger than the largest positive number found so far.
    if (setNums.has(-num) && num > 0 && num > lp) lp = num; 
    // If the conditions are met, update the largest positive number found so far.
  }

  return lp; // Return the largest positive number found, or -1 if no such number exists.
};
console.log(findMaxK(nums))

File: 2441. Largest Positive Integer That Exists With Its Negative.py
#PYTHON
#* 2441. Largest Positive Integer That Exists With Its Negative
nums = [-1,10,6,7,-7,1]
# @param {Integer[]} nums
# @return {Integer}
#* CONCISE
def findMaxK(nums):
    return max((n for n in set(nums) if n > 0 and -n in nums), default=-1)

print(findMaxK(nums))

#*SOL
def findMaxK(nums):
    uniqueNums = set(nums)
    lp = -1

    for num in uniqueNums:
        if -num in uniqueNums and num > 0:
            lp = max(lp, num)
    return lp

print(findMaxK(nums))

#* SOL
def findMaxK(nums):
    while(True):
        num=max(nums)
        if(-num in nums):
            return num
        else:
            nums.remove(num)
        if(len(nums)==0):
            return -1
print(findMaxK(nums))

#* COMMENTS
def findMaxK(nums):
    unique_nums = set(nums)  # Create a set to store unique numbers from the input list.
    largest_positive = -1  # Initialize the largest positive number found so far to -1.

    # Iterate through each number in the set of unique numbers.
    for num in unique_nums:
        # Check if the set contains the negative of the current number, and if the current number is positive.
        if -num in unique_nums and num > 0:
            # If the conditions are met, update the largest positive number found so far.
            largest_positive = max(largest_positive, num)

    # Return the largest positive number found, or -1 if no such number exists.
    return largest_positive if largest_positive != -1 else -1

print(findMaxK(nums))
